File: java/Location.java

public class Location {
	private int x, y;
	
	public Location(int x, int y) throws Exception {
		this.x = x;
		this.y = y;
	}
	
	public int[] get() {
		int[] xy = new int[2];
		xy[0] = getX();
		xy[1] = getY();
		return xy;
	}
	
	public void set(int x, int y) throws Exception {
		setX(x);
		setY(y);
	}
	
	public void moveWith(int dx, int dy) throws Exception {
		setX(getX() + dx);
		setY(getY() + dy);
	}
	
	public int getX() {
		return x;
	}
	public void setX(int x) throws Exception { // XXX why must it throw ?!!
		this.x = x;
	}
	public int getY() {
		return y;
	}
	public void setY(int y) throws Exception { // XXX why must it throw ?!!
		this.y = y;
	}
}

File: scala/Location.scala
class Location(val x: Int, val y: Int) {
    def moveWith(dx: Int, dy: Int): Location = new Location(x + dx, y + dy)
    def moveWith(delta: (Int, Int)): Location = moveWith(delta._1, delta._2)

    override def toString = x + " " + y
}

class LocationInPlateau(x: Int, y: Int, plateau: Plateau) extends Location(x, y) {
    if (! (plateau containsX x))
        throw new Exception("x is not in plateau")
    if (! (plateau containsY y))
        throw new Exception("y is not in plateau")
}